File: maven-root/mtsa/src/main/java/RobotProtocol/src/robotProtocol/messages/SetMotorSpeedMessage.cpp
/*
 * SetMotorSpeedMessage.cpp
 *
 *  Created on: Apr 5, 2013
 *      Author: mariano
 */

#include "SetMotorSpeedMessage.h"

namespace RobotProtocol {

SetMotorSpeedMessage SetMotorSpeedMessage::staticMessage = SetMotorSpeedMessage(0,0,0,0);

Message *SetMotorSpeedMessage::deserialize(char *serializedMessage){
	SetMotorSpeedMessage::staticMessage.initialize(serializedMessage);
	return &(SetMotorSpeedMessage::staticMessage);
}

SetMotorSpeedMessage::SetMotorSpeedMessage(int to, int from, int leftSpeed, int rightSpeed)
:Message(MESSAGE_ID, 1, to, from) {
	this->leftSpeed		= leftSpeed;
	this->rightSpeed	= rightSpeed;
}

SetMotorSpeedMessage::SetMotorSpeedMessage(char *serializedMessage){
	initialize(serializedMessage);
}

SetMotorSpeedMessage::~SetMotorSpeedMessage() {
}

int SetMotorSpeedMessage::getMessageLength(){
	return Message::getMessageLength() + LEFT_FIELD_LENGTH + RIGHT_FIELD_LENGTH;
}

int SetMotorSpeedMessage::getLeftSpeed(){
	return leftSpeed;
}

int SetMotorSpeedMessage::getRightSpeed(){
	return rightSpeed;
}

void SetMotorSpeedMessage::readData(char **serializedMessageIndex){
	char *serializedMessage	= *serializedMessageIndex;
	char leftField[LEFT_FIELD_LENGTH + 1];
	char rightField[RIGHT_FIELD_LENGTH + 1];
	int index = 0;
	//int messageLength = (int)strtol(*serializedMessageIndex, &next, LENGTH_FIELD_LENGTH);
	leftField[LEFT_FIELD_LENGTH] = rightField[RIGHT_FIELD_LENGTH] = '\0';
	memcpy(leftField, &(serializedMessage[index]), LEFT_FIELD_LENGTH);
	leftSpeed = (int)strtol(leftField, NULL, 10);
	index += LEFT_FIELD_LENGTH;
	memcpy(rightField, &(serializedMessage[index]), RIGHT_FIELD_LENGTH);
	rightSpeed = (int)strtol(rightField, NULL, 10);
	index += RIGHT_FIELD_LENGTH;
	(*serializedMessageIndex) += index;
}

void SetMotorSpeedMessage::writeData(char **serializedMessageIndex){
	sprintf(*serializedMessageIndex,"%04d%04d", leftSpeed, rightSpeed);
	(*serializedMessageIndex) = (*serializedMessageIndex) + LEFT_FIELD_LENGTH + RIGHT_FIELD_LENGTH;
}

} /* namespace RobotProtocol */

File: maven-root/mtsa/src/main/java/MTSAEnactment/ar/uba/dc/lafhis/enactment/robot/N6/robotProtocol/SetMotorSpeedMessage.java
package MTSAEnactment.ar.uba.dc.lafhis.enactment.robot.N6.robotProtocol;

public class SetMotorSpeedMessage extends Message {	
	public static int MESSAGE_ID				= 201;

	public static int LEFT_FIELD_LENGTH			= 4;
	public static int RIGHT_FIELD_LENGTH		= 4;	
	
	private int leftSpeed;
	
	private int rightSpeed;
	
	public int getLeftSpeed(){
		return leftSpeed;
	}
	
	public int getRightSpeed(){
		return rightSpeed;
	}
	
	public SetMotorSpeedMessage(int to, int from, int leftSpeed, int rightSpeed){
		super(MESSAGE_ID, 1, to, from);
		this.leftSpeed		= leftSpeed;
		this.rightSpeed		= rightSpeed;
	}
	
	public SetMotorSpeedMessage(String serializedMessage){
		super(serializedMessage);
	}
	
	@Override
	public int getMessageLength() {
		return super.getMessageLength() + LEFT_FIELD_LENGTH + RIGHT_FIELD_LENGTH;
	}
	
	
	@Override
	protected void readData(String inputString){
		int startIndex		= 0;
		int endIndex		= LEFT_FIELD_LENGTH;
		leftSpeed			= Integer.parseInt(inputString.substring(startIndex, endIndex),10);
		startIndex			= endIndex;
		endIndex			= startIndex + RIGHT_FIELD_LENGTH;
		rightSpeed			= Integer.parseInt(inputString.substring(startIndex, endIndex),10);
	}

	@Override
	protected String writeData(){
		return String.format("%04d%04d", leftSpeed, rightSpeed);
	}

	@Override
	protected boolean hasReply() {
		return true;
	}	
}
